client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java

```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Override
+    public void handle(final CommandContext originalCommandContext) {
+        final Command originalCommand = originalCommandContext.getCommand();
+        if (!originalCommand.isValid()) {
+            originalCommandContext.reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "malformed command message"));
+        }
+        final String tenantId = originalCommand.getTenant();
+        final String originalDeviceId = originalCommand.getDeviceId();
+        // determine last used gateway device id
+        LOG.trace("determine 'via' device to use for received command [{}]", originalCommand);
+        final Future<String> lastViaDeviceIdFuture = gatewayMapper.getMappedGatewayDevice(tenantId, originalDeviceId);
+
+        lastViaDeviceIdFuture.setHandler(deviceIdFutureResult -> {
+            if (deviceIdFutureResult.succeeded()) {
+                final String lastViaDeviceId = deviceIdFutureResult.result();
+                if (lastViaDeviceId != null) {
+                    LOG.trace("determined 'via' device {} for device {}", lastViaDeviceId, originalDeviceId);
+                    final CommandContext commandContext;
+                    if (lastViaDeviceId.equals(originalDeviceId)) {
+                        commandContext = originalCommandContext;
+                    } else {
+                        originalCommandContext.getCurrentSpan().log("determined 'via' device " + lastViaDeviceId);
+                        if (!originalCommand.isOneWay()) {
+                            originalCommand.getCommandMessage().setReplyTo(String.format("%s/%s/%s",
+                                    CommandConstants.COMMAND_ENDPOINT, tenantId, originalCommand.getReplyToId()));
+                        }
+                        final Command command = Command.from(originalCommand.getCommandMessage(), tenantId, lastViaDeviceId);
+                        commandContext = CommandContext.from(command, originalCommandContext.getDelivery(), originalCommandContext.getReceiver(), originalCommandContext.getCurrentSpan());
+                    }
+                    nextCommandHandler.handle(commandContext);
+                } else {
+                    // lastViaDeviceId is null - device hasn't connected yet
+                    LOG.error("last-via for device {} is not set", originalDeviceId);
+                    TracingHelper.logError(originalCommandContext.getCurrentSpan(), "last-via for device " + originalDeviceId + " is not set");
+                    originalCommandContext.release();
+                }
+            } else {
+                // failed to get last-via
+                LOG.error("error getting last-via for device {}", originalDeviceId, deviceIdFutureResult.cause());
+                TracingHelper.logError(originalCommandContext.getCurrentSpan(),
+                        new Exception("error getting last-via for device", deviceIdFutureResult.cause()));
+                originalCommandContext.release();
```

Comment:
I think we also should re-issue some credit so that the application can send another command, don't you think?

Reply:
See above, credit-handling being automatic on that link.